Update privacy and security landing pages (#27268)
mdn/content: files/en-us/web/security/index.md

- Put companies at a competitive disadvantage by sharing their information with competitors.
- Disable or hijack their services, again causing serious problems with their operation.
- Put their customer's [privacy](/en-US/docs/Web/Privacy) at risk, making them vulnerable to profiling, targetting, loss of data, identity theft, or even financial loss.

Modern browsers already have several features to protect users' security on the web, but developers also need to employ best practices and code carefully to ensure that their websites are secure. Even simple bugs in your code can result in vulnerabilities that bad people can exploit to steal data and get control of services they shouldn't have control of.

This set of articles provides an introduction to web security, including information that may help you understand what can make websites vulnerable, and how to secure them.

## The relationship between security and privacy

Security and privacy are closely-related topics, but they are distinct, and it is worth knowing the difference betwene the two, and how they relate.

> - **Security** is the act of keeping private data and systems protected against unauthorized access. This includes both company (internal) data, and user and partner (external) data.
> 
> - **Privacy** refers to the act of giving users the right to control how their data is collected, stored, and used, and not using it irresponsibly. For example, developers should make it very clear to users of their sites what data they are collecting, who else it will be shared with, and what they will use it for. Users must be given a chance to consent to your privacy policy, be able to see what data you have stored, and delete their data if they no longer wish you to have it.
> 
> Good security is essential for good privacy — you could go over to our [Privacy on the web](/en-US/docs/Web/Privacy) guide and follow all the advice listed there, but acting with integrity and having a solid privacy policy are futile if your site is not secure and attackers can just steal data anyway.

Reviewer: ```suggestion
- **Security** is the act of keeping private data and systems protected against unauthorized access. This includes both company (internal) data and user and partner (external) data.

- **Privacy** refers to the act of giving users the right to control how their data is collected, stored, and used, and not using it irresponsibly. For example, you should clearly communicate to users of your sites the data you are collecting, with whom it will be shared, and how it will be used. Users must be given a chance to consent to your privacy policy, have access to their data you have stored, and delete their data if they no longer wish you to have it.

Good security is very essential for good privacy. You could follow all the advice listed in our [Privacy on the web](/en-US/docs/Web/Privacy) guide, but acting with integrity and having a solid privacy policy are futile if your site is not secure and attackers can just steal data anyway.
```
- swapped "developers" with "you" to maintain consistency across sentences

## What do browsers do automatically?

Web browsers have a strict security model that enforces a good level of fundamental content, connection, and transport security. This section looks at the basics.

> ### Same-origin policy

Reviewer: should we rename this to 

```suggestion
### Same-origin policy and CORS
```

[Same-origin policy](/en-US/docs/Web/Security/Same-origin_policy) is a fundamental security mechanism of the web that restricts how a document or script loaded from one {{Glossary("origin")}} can interact with a resource from another origin. It helps isolate potentially malicious documents, reducing possible attack vectors.

> In general, documents from one origin cannot make requests to other origins. This makes sense, as you don't want sites to be able to interfere with one another and access things they shouldn't do. It does make sense to relax this restriction in some circumstances — you may have a couple of sites that interact that you want to request resources from one another, for example using [`fetch()`](/en-US/docs/Web/API/fetch).

Reviewer: "In general, documents from one origin cannot make requests to other origins."
Can we add an example for this?

```suggestion
In general, documents from one origin cannot make requests to other origins. This makes sense because you don't want sites to be able to interfere with one another and access things they shouldn't. It does make sense to relax this restriction in some circumstances; for example, you might have multiple sites that interact with each other, and you may want these sites to request resources from one another, such as using [`fetch()`](/en-US/docs/Web/API/fetch).
```

Author: The `fetch()` thing mentioned at the end of the paragraph is a pretty good example.

> This can be permitted using [Cross-Origin Resource Sharing (CORS)](/en-US/docs/Web/HTTP/CORS), an HTTP-header based mechanism that allows a server to indicate any origins (domain, scheme, or port) other than its own from which a browser should permit loading resources.

Reviewer: ```suggestion
This can be permitted using [Cross-Origin Resource Sharing (CORS)](/en-US/docs/Web/HTTP/CORS), an HTTP-header-based mechanism that allows a server to indicate any origins (domain, scheme, or port) other than its own from which a browser should permit loading resources.
```

- [How to fix a website with blocked mixed content](/en-US/docs/Web/Security/Mixed_content/How_to_fix_website_with_mixed_content)
  - : If your website delivers HTTPS pages, all [active mixed content](/en-US/docs/Web/Security/Mixed_content#mixed_active_content) delivered via HTTP on these pages will be blocked by default. Consequently, your website may appear broken to users (if iframes or plugins don't load, etc.). [Passive mixed content](/en-US/docs/Web/Security/Mixed_content#mixed_passivedisplay_content) is displayed by default, but users can set a preference to block this type of content, as well. This page explains what you should be aware of as a web developer.
- [Weak signature algorithms](/en-US/docs/Web/Security/Weak_Signature_Algorithm)
  - : The strength of the hash algorithm used in {{Glossary("Signature/Security", "signing")}} a {{Glossary("digital certificate")}} is a critical element of the security of the certificate. This article provides some information about signature algorithms known to be weak, so you can avoid them when appropriate.

### Secure contexts and feature permissions

> Browsers control usage of several "powerful features" in a couple of different ways. By "powerful features", we mean things like a site generating system notifications, using a user's web cam to get access to a media stream, manipulating the system GPU, or using web payments. If a site could just use the APIs that control such features without restriction, malicious developers could do all sorts of things, ranging from annoying users and clogging up their browser/system to create {{glossary("denial of service", "Denial of Service")}} (DoS) attacks, to stealing data and money.

Author: I've implemented these changes, sort of, but not in quite the same way you did here - "clogging up their browser/system" and "creating {{glossary("denial of service", "Denial of Service")}} (DoS) attacks" is the same point, so splitting these bits up wasn't correct.

> Browsers control usage of several "powerful features" in a couple of different ways. By "powerful features", we mean things like a site generating system notifications, using a user's web cam to get access to a media stream, manipulating the system GPU, or using web payments. If a site could just use the APIs that control such features without restriction, malicious developers could do all sorts of things, ranging from annoying users and clogging up their browser/system to create {{glossary("denial of service", "Denial of Service")}} (DoS) attacks, to stealing data and money.

Reviewer: Got it. Thanks

In addition, many other features can only be used in response to a user action such as clicking a button, meaning that they need to be invoked from inside an apropriate event handler. This is often called **transient activation**. See [Features gated by user activation](/en-US/docs/Web/Security/User_activation) for more information.

## What should client-side developers do?

There are many aspects of web security that need to be thought about on the server- and client-side. The below sections mainly focus on client-side security considerations. You can find a useful summary of security from a server-side perspective that also includes descriptions of common attacks to watch out for at [Website security](/en-US/docs/Learn/Server-side/First_steps/Website_security), which is part of our [Server-side website programming](/en-US/docs/Learn/Server-side) learning module.

> ## Store client-side data responsibly

Author: Logically, yes, probably. However, that would leave the right-hand TOC looking a bit empty, and I thought it would be good to have the different sections listed there, hence me leaving the heading levels like this

This is largely concerned with cutting down on [cookie usage](/en-US/docs/Web/HTTP/Cookies), and being careful what data you store in them — traditionally web developers have used cookies to store all kinds of data, for all kinds of purposes, and it has been easy for attackers to exploit this. As a result, browsers have started to limit what you can do with cross-site cookies, with the aim of removing access to them altogether in the future.

> It is a good idea to prepare for this, by limiting the amount of tracking activities you rely on, and/or implementing desired information persistence in other ways. For example:

Author: I've done this one, but in a slightly different way. It is really more about preparing for complete removal of cross-site cookies, than limiting usage. The latter has already happened.

> - Use an alternative client-side storage mechanism such as [Web Storage](/en-US/docs/Web/API/Web_Storage_API) to persist data. This does have the downside that the data is stored per-origin, so it can't be shared.

Reviewer 2: it might be worth it to point out the different types of web storage and their security implications, we usually prefer developers to use sessionStorage since data stored in localStorage is persistent between sessions.

Author: Good idea. In the next commit, I've added to this bullet to point out the two different variants and the difference between them, and recommend `sessionStorage`.

> - Use an alternative client-side storage mechanism such as [Web Storage](/en-US/docs/Web/API/Web_Storage_API) to persist data. This does have the downside that the data is stored per-origin, so it can't be shared.

Author: "such as using the" sounded a bit awkward. I've done this one slightly differently.

- Use a technology such as [IFrame credentialless](/en-US/docs/Web/Security/IFrame_credentialless), [First-party sets](https://developer.chrome.com/docs/privacy-sandbox/first-party-sets/) and/or the [Storage access API](/en-US/docs/Web/API/Storage_Access_API) to allow your sites to opt-in to using cross-site cookies in a safe, controlled way, or block them altogether. These currently have limited browser support.
- Use a server-side solution for data persistence.

See our [privacy guide](/en-US/docs/Web/Privacy) and particularly [Cut down on tracking cookies](/en-US/docs/Web/Privacy#cut_down_on_tracking_cookies) for more context around this.

> ## Protect user identity and manage logins

Author: As mentioned earlier, I want to keep these as h2s so they are more visible.

When implementing a secure solution that involves data collection — particularly if it is sensitive data such as log-in credentials — it makes sense to use a reputable solution from a well-respected provider. For example, any respectable server-side framework will have built-in features to protect against common vulnerabilities. You could also consider using a specialized product for your purpose — for example an identity provider solution, or a secure online survey provider.

If you want to roll your own solution for collecting user data, make sure you understand what you are doing. Hire an experienced server-side developer and/or security engineer to implement the system, and ensure it is tested throroughly. Use multi-factor authentication (MFA) to provide better protection. Consider using a dedicated API such as [Web Authentication](/en-US/docs/Web/API/Web_Authentication_API) or [Federated Credential Management](/en-US/docs/Web/API/FedCM_API) to streamline the client-side of the app.

Other tips for secure logins:

> - When collecting user login information, enforce strong passwords so that your user's account details cannot be easily guessed. Weak passwords are one of the main causes of security breaches. In addition, encourage your users to use a password manager so that they can use more complex passwords, don't need to worry about remembering them, and won't create a security risk by writing them down. See also [Insecure passwords](/en-US/docs/Web/Security/Insecure_passwords).

Author: This should have been "our article on"; I've done this in the next commit.

> - When collecting user login information, enforce strong passwords so that your user's account details cannot be easily guessed. Weak passwords are one of the main causes of security breaches. In addition, encourage your users to use a password manager so that they can use more complex passwords, don't need to worry about remembering them, and won't create a security risk by writing them down. See also [Insecure passwords](/en-US/docs/Web/Security/Insecure_passwords).

Reviewer: Oops, yeah. Sorry about that and thanks for catching it.

> - The flipside of this is educating your users against phishing, which is the act of sending a user an email, SMS, etc. containing a link to a site that looks like a site they use every day, but isn't. This is combined with a message designed to trick them into entering their username and password so it can be stolen and then used by an attacker. Some phishing sites can be very sophisticated and hard to distinguish from the real thing. You should educate your users to not trust random links in emails and SMS messages. If they receive a message along the lines of "urgent, you need to log in now to resolve an issue", they should go to the site directly in a clean tab and try logging in rather than following the link. Or they could phone or email you to discuss it.

Reviewer: - "The flipside of this is educating": can we clarify "this" here
- breaking up the sentence to have the phishing definition in an independent sentence

```suggestion
- The flipside of this is educating your users against phishing. The technique of phishing is used by attackers where they send a user an email, SMS, etc. that contains a link to a site. When users click on the link, it takes them to a site that looks like a site they use every day but isn't. The phishing link is combined with a message designed to trick users into entering their username and password with the intention of stealing and using this data for malicious purposes. Some phishing sites can be very sophisticated and hard to distinguish from the real website. You should, therefore, educate your users to not trust random links in emails and text messages. If users receive a message along the lines of "Urgent, you need to log in now to resolve an issue", they should go to the site directly in a new tab and try logging into the website directly rather than clicking the link in the email/message. Or they could phone or email you to discuss the message they received.
```

Author: I've implemented quite a lot of these changes but done it in a slightly different way. The two key issues you brought up have been fixed.

- Protect against brute force attacks on login pages with [rate limiting](https://www.cloudflare.com/en-gb/learning/bots/what-is-rate-limiting/), account lockouts after a certain number of unsuccessful attempts, and [CAPTCHA challenges](https://en.wikipedia.org/wiki/CAPTCHA).

> - Manage user login sessions with unique [session IDs](https://en.wikipedia.org/wiki/Session_ID), and automatically log out after periods of inactivity.

Reviewer: ```suggestion
- Manage user login sessions with unique [session IDs](https://en.wikipedia.org/wiki/Session_ID) and automatically logging out users after periods of inactivity.
```

Author: I've just added "users" after "log out", in this case. Your rewording sounded at odds with the start of the sentence.

> As a general rule, you shouldn't include sensitive data in URLs — if a third party intercepts the URL (for example via the {{httpheader("Referer")}} header), they could steal that information. Use `POST` requests rather than `GET` requests to avoid this. [Referer header policy: privacy and security concerns](/en-US/docs/Web/Security/Referer_header:_privacy_and_security_concerns) describes in more detail the privacy and security risks associated with the [Referer HTTP header](/en-US/docs/Web/HTTP/Headers/Referer), and offers advice on mitigating those risks.

Reviewer 2: another common method of leaking query string parameters in GET requests is indexing and archiving services

Author: @fkiriakos07 have you got a link to share that provides more context around this? I tried searching, but couldn't really find anything useful, so I wasn't sure what to write here.

Reviewer 2: I couldn't find lot of interesting resources too, maybe here: https://owasp.org/www-community/vulnerabilities/Information_exposure_through_query_strings_in_url.

Maybe we can point out that in addition to leaking the parameters in the referrer header, the GET URLs can also be indexed by public web crawlers, HTTP proxies and archiving tools such as the internet archive, https://web.archive.org/

Author: Ah, now I understand, thanks! I've put this stuff under a new heading and added information about the archiving risks.

> As a general rule, you shouldn't include sensitive data in URLs — if a third party intercepts the URL (for example via the {{httpheader("Referer")}} header), they could steal that information. Use `POST` requests rather than `GET` requests to avoid this. [Referer header policy: privacy and security concerns](/en-US/docs/Web/Security/Referer_header:_privacy_and_security_concerns) describes in more detail the privacy and security risks associated with the [Referer HTTP header](/en-US/docs/Web/HTTP/Headers/Referer), and offers advice on mitigating those risks.

Reviewer: ```suggestion
As a general rule, you shouldn't include sensitive data in URLs; if a third party intercepts the URL (for example via the {{httpheader("Referer")}} header), they could steal that information. Use `POST` requests rather than `GET` requests to avoid this. Our article on [Referer header policy: Privacy and security concerns](/en-US/docs/Web/Security/Referer_header:_privacy_and_security_concerns) describes in more detail the privacy and security risks associated with the [Referer HTTP header](/en-US/docs/Web/HTTP/Headers/Referer) and offers advice on mitigating those risks.
```
that needs a fix in the page title as well

Author: I've done all of this, except that I've used a color rather than a semi-colon in the first sentence - the second article is really a lead-on from the first article, not a separate point.

> > **Note**: Steering away from transmitting sensitive data in URLs via `GET` requests can help protect against {{glossary("CSRF", "cross-site request forgery")}} and [replay attacks](https://en.wikipedia.org/wiki/Replay_attack).

Reviewer: for some reason, this note is not rendering in a blue box

Author: Ah, it was because the colon was outside the asterisks; it's pretty picky about the syntax order. Fixed now!

Reviewer: Good catch, thanks!

## Enforce usage policies

> Consider using tools like [Content Security Policy](/en-US/docs/Web/HTTP/CSP) (CSP) and [Permissions Policy](/en-US/docs/Web/HTTP/Permissions_Policy) to enforce a set of feature and resource usage on your site that makes it harder to introduce vulnerabilities.

Reviewer: ```suggestion
Consider using tools like [Content Security Policy](/en-US/docs/Web/HTTP/CSP) (CSP) and [Permissions policy](/en-US/docs/Web/HTTP/Permissions_Policy) to enforce a set of feature and resource usage on your site that makes it harder to introduce vulnerabilities.
```

> CSP allows you to add a layer of security by for example only allowing images or scripts to be loaded from specific trusted origins. This helps to detect and mitigate certain types of attacks, including Cross-Site Scripting ({{Glossary("Cross-site_scripting", "XSS")}}) and data injection attacks. These attacks are used for everything from data theft to site defacement to the distribution of malware.

Reviewer: ```suggestion
CSP allows you to add a layer of security by, for example, allowing images or scripts to be loaded only from specific trusted origins. This helps to detect and mitigate certain types of attacks, including Cross-Site Scripting ({{Glossary("Cross-site_scripting", "XSS")}}) and data injection attacks. These attacks involve a range of malicious activities, including data theft, site defacement, and distribution of malware.
```

> Permissions policy works in a similar way, except that it is more concerned with allowing or blocking access to specific "powerful features" (as mentioned earlier).

Reviewer: ```suggestion
Permissions policy works in a similar way, except that it is more concerned with allowing or blocking access to specific "powerful features" ([as mentioned earlier](#secure_contexts_and_feature_permissions)).
```

> > **Note:** Such policies are very useful to help keep sites secure, especially when you are using a lot of third party code on your site. Be careful however — if you block usage of a feature that a third party script relies on to work, you may end up breaking your site's functionality.

Reviewer: ```suggestion
> **Note:** Such policies are very useful to help keep sites secure, especially when you are using a lot of third party code on your site. However, keep in mind that if you block usage of a feature that a third-party script relies on to work, you may end up breaking your site's functionality.
```

## Maintain data integrity

> Following on from the previous section, when you do allow feature and resource usage on your site, you should try to ensure that resources have not been tempered with.

Reviewer 2: ```suggestion
Following on from the previous section, when you do allow feature and resource usage on your site, you should try to ensure that resources have not been tampered with.
```

Related topics:

- [Subresource integrity](/en-US/docs/Web/Security/Subresource_Integrity)
  - : **Subresource Integrity** (SRI) is a security feature that enables browsers to verify that resources they fetch (for example, from a {{Glossary("CDN")}}) are delivered without unexpected manipulation. It works by allowing you to provide a cryptographic hash that a fetched resource must match.
- [HTTP Access-Control-Allow-Origin](/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin)
  - : The **`Access-Control-Allow-Origin`** response header indicates whether the response can be shared with requesting code from the given {{glossary("origin")}}.
- [HTTP X-Content-Type-Options](/en-US/docs/Web/HTTP/Headers/X-Content-Type-Options)
  - : The **`X-Content-Type-Options`** response HTTP header is a marker used by the server to indicate that the [MIME types](/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types) advertised in the {{HTTPHeader("Content-Type")}} headers should not be changed and be followed. This is a way to opt out of [MIME type sniffing](/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types#mime_sniffing), or, in other words, to say that the MIME types are deliberately configured.

> ## Sanitize form input

Reviewer 2: a very important method to protect against injections is to use output encoding (or escaping) when displaying user input on the application. I think it's worth adding a section about it, https://owasp.org/www-project-proactive-controls/v3/en/c4-encode-escape-data

To protect against this, you should thoroughly sanitize data entered into your forms.

First of all, you should implement client-side validation to inform users when they have entered data in the wrong format. You can do this using built-in HTML form validation features, or you can write your own validation code. See [Client-side form validation](/en-US/docs/Learn/Forms/Form_validation) for more information.

However, you can't rely on client-side validation alone for security. It is a useful user experience enhancement for your users because it gives them instant validation feedback without having to wait for a round trip to the server. However, it is too easy for a malicious party to bypass (for example they could turn off JavaScript in the browser to bypass JavaScript-based validation), so it should be combined with server-side validation.

Any reputable server-side framework will provide functionality for validating form submissions. In addition, a common best practice is to escape any special characters that form part of executable syntax, thereby making any entered code no longer executable and treated as plain text.

## Protect against clickjacking

> In [clickjacking](/en-US/docs/Glossary/Clickjacking), a user is fooled into clicking on a UI element that performs some action other than what the user expects. This can be a risk inherent in embedded third party content (make sure you trust what is being embedded into your site), and it can also be combined with phishing.

Reviewer: ```suggestion
In [clickjacking](/en-US/docs/Glossary/Clickjacking), a user is fooled into clicking on a UI element that performs an action different from what the user expects. This can be a risk inherent in embedded third-party content (make sure you trust what is being embedded into your site), and it can also be combined with phishing.
```

The following features can help guard against clickjacking:

- [HTTP X-Frame-Options](/en-US/docs/Web/HTTP/Headers/X-Frame-Options)
  - : The **`X-Frame-Options`** [HTTP](/en-US/docs/Web/HTTP) response header can be used to indicate whether a browser should be allowed to render a page in a [`<frame>`](/en-US/docs/Web/HTML/Element/frame), [`<iframe>`](/en-US/docs/Web/HTML/Element/iframe), [`<embed>`](/en-US/docs/Web/HTML/Element/embed) or [`<object>`](/en-US/docs/Web/HTML/Element/object). Sites can use this to avoid [clickjacking](/en-US/docs/Glossary/Clickjacking) attacks, by ensuring that their content is not embedded into other sites.
- [CSP: frame-ancestors](/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/frame-ancestors)
  - : The HTTP {{HTTPHeader("Content-Security-Policy")}} (CSP) **`frame-ancestors`** directive specifies valid parents that may embed a page using {{HTMLElement("frame")}}, {{HTMLElement("iframe")}}, {{HTMLElement("object")}}, or {{HTMLElement("embed")}}.

## See also

- [Privacy on the web](/en-US/docs/Web/Privacy)
- [Website security](/en-US/docs/Learn/Server-side/First_steps/Website_security)

> - [Security Blog](https://blog.mozilla.org/security/)

Reviewer: ```suggestion
- [Mozilla security blog](https://blog.mozilla.org/security/)
```

If you want to roll your own solution for collecting user data, make sure you understand all aspects and requirements. Hire an experienced server-side developer and/or security engineer to implement the system, and ensure it is tested thoroughly. Use multi-factor authentication (MFA) to provide better protection. Consider using a dedicated API such as [Web Authentication](/en-US/docs/Web/API/Web_Authentication_API) or [Federated Credential Management](/en-US/docs/Web/API/FedCM_API) to streamline the client-side of the app.

Here are some other tips for providing secure logins:

- When collecting user login information, enforce strong passwords so that your user's account details cannot be easily guessed. Weak passwords are one of the main causes of security breaches. In addition, encourage your users to use a password manager so that they can use more complex passwords, don't need to worry about remembering them, and won't create a security risk by writing them down. See also our article on [Insecure passwords](/en-US/docs/Web/Security/Insecure_passwords).

> - You should also educate your users about **phishing**. Phishing is the act of sending a user a message (for example an email or SMS) containing a link to a site that looks like a site they use every day, but isn't. This is accompanied by a message designed to trick them into entering their username and password on the site so it can be stolen and then used by an attacker for malicious purposes.

Reviewer: ```suggestion
- You should also educate your users about **phishing**. Phishing is the act of sending a message to a user (for example, an email or an SMS) containing a link to a site that looks like a site they use every day but isn't. The link is accompanied by a message designed to trick users into entering their username and password on the site so it can be stolen and then used by an attacker for malicious purposes.
```

>   > **Note:** Some phishing sites can be very sophisticated and hard to distinguish from the real thing. You should therefore educate your users to not trust random links in emails and SMS messages. If they receive a message along the lines of "Urgent, you need to log in now to resolve an issue", they should go to the site directly in a new tab and try logging in directly rather than clicking the link in the message. Or they could phone or email you to discuss it.

Reviewer: ```suggestion
  > **Note:** Some phishing sites can be very sophisticated and hard to distinguish from the real website. You should therefore educate your users to not trust random links in emails and SMS messages. If they receive a message along the lines of "Urgent, you need to log in now to resolve an issue", they should go to the site directly in a new tab and try logging in directly rather than clicking the link in the message. Or they could phone or email you to discuss the message they received.
```
